core/src/main/java/io/parsingdata/metal/token/Tie.java

```diff
@@ -57,28 +56,28 @@ public Tie(final String name, final Token token, final ValueExpression dataExpre
     }
 
     @Override
-    protected ParseResult parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
+    protected Optional<Environment> parseImpl(final String scope, final Environment environment, final Encoding encoding) throws IOException {
         final ImmutableList<Optional<Value>> dataResult = dataExpression.eval(environment.order, encoding);
         if (dataResult.isEmpty()) {
-            return failure(environment);
+            return failure();
         }
-        final ParseResult result = iterate(scope, dataResult, 0, environment.addBranch(this), encoding);
-        if (result.succeeded) {
-            return success(new Environment(result.environment.closeBranch().order, environment.source, environment.offset, environment.callbacks));
+        final Optional<Environment> result = iterate(scope, dataResult, 0, environment.addBranch(this), encoding);
+        if (result.isPresent()) {
+            return success(new Environment(result.get().closeBranch().order, environment.source, environment.offset, environment.callbacks));
```

Comment:
~~Can we just pass `result.get().closeBranch()`, here instead of reconstructing an entire `new Environment`?~~

Nevermind. But could we do `result.get().closeBranch().source(environment.source).seek(environment.offset)`?